src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -140,38 +147,119 @@ internal void Freeze()
             _orderedSynthesizedMethods = _synthesizedMethods.OrderBy(kvp => kvp.Key).Select(kvp => kvp.Value).AsImmutable();
 
             // Sort proxy types.
-            _orderedProxyTypes = _proxyTypes.OrderBy(kvp => kvp.Key).Select(kvp => kvp.Value).AsImmutable();
+            _orderedProxyTypes = _proxyTypes.OrderBy(kvp => kvp.Key.Size).Select(kvp => kvp.Value).AsImmutable();
         }
 
         private bool IsFrozen => _frozen != 0;
 
-        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data)
+        /// <summary>
+        /// Gets a field that can be used to cache an array allocated to store data from a corresponding <see cref="CreateDataField"/> call.
+        /// </summary>
+        /// <param name="data">The data that will be used to initialize the field.</param>
+        /// <param name="arrayType">The type of the field, e.g. int[].</param>
+        /// <param name="emitContext">The emit context to use with the array type to extract its element type.</param>
+        /// <returns>The field to use to cache an array for this data and alignment.</returns>
+        internal Cci.IFieldReference CreateArrayCachingField(ImmutableArray<byte> data, Cci.IArrayTypeReference arrayType, EmitContext emitContext)
+        {
+            Debug.Assert(!IsFrozen);
+
+            // Get the SpecialType for the array's element type.
+            INamedTypeSymbol? symbol = arrayType.GetElementType(emitContext).GetInternalSymbol()?.GetISymbol() as INamedTypeSymbol;
```

Comment:
>arrayType.GetElementType(emitContext).GetInternalSymbol()?.GetISymbol() [](http://example.com/codeflow?start=39&length=71)

It looks like we are mixing three different layers in this expression: Cci, internal symbol interfaces, and public symbol interfaces. We should stick with just Cci layer and not make assumptions that objects we get here implement any other layer.

Reply:
So use Cci.PrimitiveTypeCode instead of SpecialType?

Reply:
There is a ```ITypeReference.TypeCode``` API. There is a helper ```SpecialType SpecialTypes.GetTypeFromMetadataName(Microsoft.Cci.PrimitiveTypeCode typeCode)```, which can be used to go to ```SpecialType```. But in this layer, it is probably better to stick with PrimitiveTypeCode in order to keep API layers separate.